File: pyNA/src/noise_src_py/core_source.py
import pdb
import openmdao
import openmdao.api as om
import numpy as np


def core_source(source, theta, inputs: openmdao.vectors.default_vector.DefaultVector) -> np.ndarray:
	"""
	Compute core noise mean-square acoustic pressure (msap).

	:param source: pyNA component computing noise sources
	:type source: Source
	:param inputs: unscaled, dimensional input variables read via inputs[key]
	:type inputs: openmdao.vectors.default_vector.DefaultVector

	:return: msap_core
	:rtype: np.ndarray [n_t, settings['n_frequency_bands']]
	"""
	# Load options
	settings = source.options['settings']
	data = source.options['data']
	airframe = source.options['airframe']
	n_t = source.options['n_t']

	# Extract inputs
	M_0 = inputs['M_0']
	if settings['core_turbine_attenuation_method'] == 'ge':
		mdoti_c_star = inputs['mdoti_c_star']
		Tti_c_star = inputs['Tti_c_star']
		Ttj_c_star = inputs['Ttj_c_star']
		Pti_c_star = inputs['Pti_c_star']
		DTt_des_c_star = inputs['DTt_des_c_star']
		
	elif settings['core_turbine_attenuation_method'] == 'pw':
		mdoti_c_star = inputs['mdoti_c_star']
		Tti_c_star = inputs['Tti_c_star']
		Ttj_c_star = inputs['Ttj_c_star']
		Pti_c_star = inputs['Pti_c_star']
		rho_te_c_star = inputs['rho_te_c_star']
		c_te_c_star = inputs['c_te_c_star']
		rho_ti_c_star = inputs['rho_ti_c_star']
		c_ti_c_star = inputs['c_ti_c_star']
	else:
		raise ValueError('Invalid method for turbine noise in core module. Specify: GE/PW.')
	
	r_s_star = settings['r_0'] / np.sqrt(settings['A_e'])
	A_c_star = 1.

	# Initialize solution
	msap_core = np.zeros((n_t, settings['n_frequency_bands']))

	# Compute core
	for i in np.arange(n_t):

		# Turbine transmission loss function
		# Source: Zorumski report 1982 part 2. Chapter 8.2 Equation 3
		if settings['core_turbine_attenuation_method'] == 'ge':
			g_TT = DTt_des_c_star[i] ** (-4)
		# Source: Hultgren, 2012: A comparison of combustor models Equation 6
		elif settings['core_turbine_attenuation_method'] == 'pw':
			zeta = (rho_te_c_star[i] * c_te_c_star[i]) / (rho_ti_c_star[i] * c_ti_c_star[i])
			g_TT = 0.8 * zeta / (1 + zeta) ** 2
		else:
			raise ValueError('Invalid method to account for turbine attenuation effects of combustor noise. Specify GE/PW.')

		# Calculate acoustic power (Pi_star)
		# Source Zorumski report 1982 part 2. Chapter 8.2 Equation 3
		Pi_star = 8.85e-7 * (mdoti_c_star[i] / A_c_star) * ((Ttj_c_star[i] - Tti_c_star[i]) / Tti_c_star[i]) ** 2 * Pti_c_star[i] ** 2 * g_TT

		# Calculate directivity function (D)
		# Take the D function as SAE ARP876E Table 18 and all other values which are not in the table from Zorumski report 1982 part 2. Chapter 8.2 Table II
		array_1 = np.linspace(0, 180, 19)
		array_2 = np.array([-0.85, -0.8, -0.75, -0.7, -0.65, -0.6, -0.53, -0.46, -0.39, -0.16, 0.08, 0.31, 0.5, 0.35, 0.12,-0.19,-0.51, -0.8, -0.9])
		D_function = np.interp(theta[i], array_1, array_2)
		D_function = 10 ** D_function

		# Calculate the spectral function (S)
		# Source Zorumski report 1982 part 2. Chapter 8.2 Equation 4
		f_p = 400 / (1 - M_0[i] * np.cos(theta[i] * np.pi / 180.))

		log10ffp = np.log10(data.f / f_p)

		# Take the S function as SAE ARP876E Table 17 and all other values which are not in the table from Zorumski report 1982 part 2. Chapter 8.2 Table III
		array_1 = np.linspace(-1.1, 1.6, 28)
		array_2 = np.array([-3.87, -3.47, -3.12, -2.72, -2.32, -1.99, -1.7, -1.41, -1.17, -0.97, -0.82, -0.72, -0.82, -0.97,
							-1.17, -1.41, -1.7, -1.99, -2.32, -2.72, -3.12, -3.47, -3.87, -4.32, -4.72, -5.22, -5.7, -6.2])
		S_function = np.interp(log10ffp, array_1, array_2)
		S_function = 10 ** S_function

		# Calculate mean-square acoustic pressure (msap)
		# Source Zorumski report 1982 part 2. Chapter 8.2 Equation 1
		msap_j = Pi_star * A_c_star / (4 * np.pi * r_s_star ** 2) * D_function * S_function / (1. - M_0[i] * np.cos(np.pi / 180. * theta[i])) ** 4

		# Multiply with number of engines
		msap_j = msap_j * airframe.n_eng

		# Normalize msap by reference pressure
		msap_core[i, :] = msap_j/settings['p_ref']**2

	return msap_core

File: pyNA/src/noise_src_jl/core_source.jl
using ReverseDiff

function core_source_ge!(spl::Array, x::Union{Array, ReverseDiff.TrackedArray}, settings, pyna_ip, af, f::Array{Float64,1})

    # x = [mdoti_c, Tti_c, Ttj_c, Pti_c, DTt_des_c, c_0, rho_0, T_0, p_0, M_0, TS, theta]
    # y = spl
    
    # normalize engine variables
    mdoti_c_star = x[1] / (x[7] * x[6] * settings["A_e"])
    Tti_c_star = x[2] / x[8]
    Ttj_c_star = x[3] / x[8]
    Pti_c_star = x[4] / x[9]
    DTt_des_c_star = x[5] / x[8]

    # Initialize solution
    r_s_star = settings["r_0"]/sqrt(settings["A_e"])
    A_c_star = 1.

    # Turbine transmission loss function
    # Source: Zorumski report 1982 part 2. Chapter 8.2 Equation 3
    g_TT = DTt_des_c_star^(-4)
    
    # Calculate acoustic power (Pi_star)
    # Source Zorumski report 1982 part 2. Chapter 8.2 Equation 3
    Pi_star = 8.85e-7 * (mdoti_c_star / A_c_star) * ((Ttj_c_star - Tti_c_star) / Tti_c_star)^2 * Pti_c_star^2 * g_TT
    
    # Calculate directivity function (D)
    D_funct = 10 ^pyna_ip.f_D_core(x[12])

    # Calculate the spectral function (S)
    # Source Zorumski report 1982 part 2. Chapter 8.2 Equation 4
    f_p = 400 / (1 - x[10] * cos.(x[12] * π / 180.))
    log10ffp = log10.(f / f_p)
    S_funct = 10 .^pyna_ip.f_S_core.(log10ffp)
      
    # # Calculate mean-square acoustic pressure (msap)
    # # Source Zorumski report 1982 part 2. Chapter 8.2 Equation 1
    # # Multiply with number of engines and normalize msap by reference pressure
    pow_level_core = af.n_eng / settings["p_ref"]^2 * Pi_star * A_c_star / (4 * π * r_s_star^2) * D_funct / (1 - x[10] * cos(π / 180. * x[12]))^4
    if settings["core_jet_suppression"] && settings["case_name"] in ["nasa_stca_standard", "stca_enginedesign_standard"] && x[11] > 0.8
        spl .+= pow_level_core * 10 ^(-2.3 / 10.) * S_funct
    else
        spl .+= pow_level_core * S_funct
    end
end


function core_source_pw!(spl::Array, x::Union{Array, ReverseDiff.TrackedArray}, settings, pyna_ip, af, f::Array{Float64, 1})
    
    # x = [mdoti_c, Tti_c, Ttj_c, Pti_c, rho_te_c, c_te_c, rho_ti_c, c_ti_c, c_0, T_0, rho_0, p_0, M_0, TS, theta]
    # y = spl

    # Normalize engine variables
    mdoti_c_star = x[1] / (x[9] * settings["A_e"] * x[11])
    Tti_c_star = x[2] / x[10]
    Ttj_c_star = x[3] / x[10]
    Pti_c_star = x[4] / x[12]
    rho_te_c_star = x[5] / x[11]
    c_te_c_star = x[6] / x[9]
    rho_ti_c_star = x[7] / x[11]
    c_ti_c_star = x[8] / x[9]

    # Initialize solution
    r_s_star = settings["r_0"]/sqrt(settings["A_e"])
    A_c_star = 1.

    # Turbine transmission loss function
    # Source: Hultgren, 2012: A comparison of combustor models Equation 6
    zeta = (rho_te_c_star*c_te_c_star) / (rho_ti_c_star*c_ti_c_star)
    g_TT = 0.8 * zeta / (1 + zeta)^2
    
    # Calculate acoustic power (Pi_star)
    # Source Zorumski report 1982 part 2. Chapter 8.2 Equation 3
    Pi_star = 8.85e-7 * (mdoti_c_star / A_c_star) * ((Ttj_c_star - Tti_c_star) / Tti_c_star)^2 * Ptj_c_star^2 * g_TT
    
    # Calculate directivity function (D)
    D_funct = 10 ^pyna_ip.f_D_core(x[15])

    # Calculate the spectral function (S)
    # Source Zorumski report 1982 part 2. Chapter 8.2 Equation 4
    f_p = 400 / (1 - x[13] * cos(x[15] * π / 180.))
    log10ffp = log10.(f / f_p)

    # Take the S function as SAE ARP876E Table 17 and all other values which are not in the table from Zorumski report 1982 part 2. Chapter 8.2 Table III
    S_funct = 10 .^pyna_ip.f_S_core.(log10ffp)

    # Calculate mean-square acoustic pressure (msap)
    # Source Zorumski report 1982 part 2. Chapter 8.2 Equation 1
    # Multiply with number of engines and normalize msap by reference pressure
    pow_level_core = af.n_eng / settings["p_ref"]^2 * Pi_star * A_c_star / (4 * π * r_s_star^2) * D_funct / (1. - x[13] * cos.(π / 180. * x[15]))^4
    if settings["core_jet_suppression"] && settings["case_name"] in ["nasa_stca_standard", "stca_enginedesign_standard"] && x[14] > 0.8
        spl .+= pow_level_core * 10 ^(-2.3 / 10.) * S_funct
    else
        spl .+= pow_level_core * S_funct
    end
end

core_source_ge_fwd! = (y,x)->core_source_ge!(y, x, settings, pyna_ip, ac, f)
core_source_pw_fwd! = (y,x)->core_source_pw!(y, x, settings, pyna_ip, ac, f)